src/tensor_ops/realize_to.rs

```diff
@@ -0,0 +1,213 @@
+use crate::{shapes::*, tensor::*};
+
+/// Changes order of dimensions/axes
+pub trait RealizeTo: HasErr + HasShape {
+    /// Realizes the concrete shape of the tensor as another compatable shape:
+    /// ```rust
+    /// # use dfdx::prelude::*;
+    /// # let dev: Cpu = Default::default();
+    /// let a: Tensor<Rank2<2, 3>, f32, _> = dev.zeros();
+    /// let a = a.realize::<(usize, usize)>().unwrap();
+    /// let a = a.realize::<Rank2<2, 3>>().unwrap();
+    /// ```
+    fn realize<Dst: Shape>(self) -> Option<Self::WithShape<Dst>>
+    where
+        Self::Shape: RealizeShapeTo<Dst>,
+    {
+        self.try_realize().unwrap()
+    }
+
+    /// Fallible version of [RealizeTo::realize].
+    fn try_realize<Dst: Shape>(self) -> Result<Option<Self::WithShape<Dst>>, Self::Err>
+    where
+        Self::Shape: RealizeShapeTo<Dst>;
+}
+
+impl<S: Shape, E: Dtype, D: DeviceStorage, T: Tape<E, D>> RealizeTo for Tensor<S, E, D, T> {
+    fn try_realize<Dst: Shape>(self) -> Result<Option<Self::WithShape<Dst>>, Self::Err>
+    where
+        Self::Shape: RealizeShapeTo<Dst>,
+    {
+        Ok(self.shape.realized().map(|dst_shape| Tensor {
+            id: self.id,
+            data: self.data,
+            strides: dst_shape.strides(),
+            shape: dst_shape,
+            device: self.device,
+            tape: self.tape,
+        }))
+    }
+}
```

Comment:
It should be fine to do something like the following unless you want to have an error type to describe the specific error:

```suggestion
impl<S: Shape, E: Dtype, D: DeviceStorage, T: Tape<E, D>> RealizeTo for Tensor<S, E, D, T> {
    fn try_realize<Dst: Shape>(self) -> Result<Result<Self::WithShape<Dst>, Self>, Self::Err>
    where
        Self::Shape: RealizeShapeTo<Dst>,
    {
        if let Some(dst_shape) = self.shape.realized() {
            Ok(Ok(Tensor {
                id: self.id,
                data: self.data,
                strides: dst_shape.strides(),
                shape: dst_shape,
                device: self.device,
                tape: self.tape,
            }))
        } else {
            Ok(Err(self))
        }
    }
}
```

Reply:
It's strange to have the ability to specify an error type and simply refer back to one that's so limiting and generic. My preference would be to have an error type for this trait, but I wasn't sure what @coreylowman's thoughts were on it.

Reply:
TBH I don't even think we need try_realize. The main point of try_* right now is to catch memory allocation errors or other device level errors (like launching cuda kernel errors). Since realize doesn't re-allocate, or really do anything at the device level, it won't ever fail like this.

Reply:
In that case, should `realize` return a `Result<_, Self>` or something similar to allow recovering from a failed call?

Reply:
Yeah I like that idea!